### ISSUE TITLE
Caching issue

### ISSUE THREAD

**Issue by ailzhang** [pytorch/xla#1808]
@dlibenzi I seem to hit some caching issue when debugging a test failure, see the repro script below. 
If I comment out the target line, everything is correct. But when it's present, the last print produce wrong result.  
```
import torch
import math
import torch_xla
import torch_xla.core.xla_model as xm

device = xm.xla_device()
dtype = torch.uint8

m1 = torch.empty(0, dtype=dtype, device=device)
range_high = 4 if dtype in (torch.int8, torch.uint8) else 10
m1 = torch.randint(1, range_high, (100, 100), dtype=dtype, device=device)

res1 = torch.pow(3, m1[4])
res2 = res1.clone().zero_()
for i in range(res2.size(0)):
    res2[i] = math.pow(3, m1[4, i])
print(res2) <--- everything is fine if commenting this line

res1 = torch.pow(3, m1[:, 4])
res2 = res1.clone().zero_()
for i in range(res2.size(0)):
    res2[i] = math.pow(3, m1[i][4])
print(m1[:, 4])
print(res2) 
```

**Comment by dlibenzi** [comment]
With this:
```Python
import torch
import math
import torch_xla
import torch_xla.core.xla_model as xm

def to_cpu(v):
  return v


def test(device, dtype):
  print('DEVICE', str(device))
  torch.manual_seed(17)
  range_high = 5
  m1 = torch.randint(1, range_high, (6, 6), dtype=dtype, device=device)
  print('m1.cpu()[:, 4]\n', to_cpu(m1)[:, 4])

  res1 = torch.pow(3, m1[4])
  print('torch.pow(3, m1[4])\n', to_cpu(res1))
  res2 = res1.clone().zero_()
  for i in range(res2.size(0)):
    res2[i] = math.pow(3, m1[4, i])
  print('res2\n', to_cpu(res2)) # <--- everything is fine if commenting this line
  # print(torch_xla._XLAC._get_xla_tensors_text([res2]))

  res1 = torch.pow(3, m1[:, 4])
  print('torch.pow(3, m1[:, 4])\n', to_cpu(res1))
  res2 = res1.clone().zero_()
  for i in range(res2.size(0)):
    res2[i] = math.pow(3, m1[i][4])
  print('res2\n', to_cpu(res2))
  # print(torch_xla._XLAC._get_xla_tensors_text([res2]))


test(xm.xla_device(), torch.uint8)
test('cpu', torch.uint8)

```

I get this:
```
DEVICE xla:0
m1.cpu()[:, 4]
 tensor([3, 1, 3, 4, 2, 2], device='xla:0', dtype=torch.uint8)
torch.pow(3, m1[4])
 tensor([27, 81, 27, 81,  9,  9], device='xla:0', dtype=torch.uint8)
res2
 tensor([27, 81, 27, 81,  9,  9], device='xla:0', dtype=torch.uint8)
torch.pow(3, m1[:, 4])
 tensor([27,  3, 27, 81,  9,  9], device='xla:0', dtype=torch.uint8)
res2
 tensor([27,  3, 27, 81,  9,  9], device='xla:0', dtype=torch.uint8)

DEVICE cpu
m1.cpu()[:, 4]
 tensor([3, 1, 3, 4, 2, 2], dtype=torch.uint8)
torch.pow(3, m1[4])
 tensor([27, 81, 27, 81,  9,  9], dtype=torch.uint8)
res2
 tensor([27, 81, 27, 81,  9,  9], dtype=torch.uint8)
torch.pow(3, m1[:, 4])
 tensor([27,  3, 27, 81,  9,  9], dtype=torch.uint8)
res2
 tensor([27,  3, 27, 81,  9,  9], dtype=torch.uint8)
```

**Comment by ailzhang** [comment]
fixed by #1829
